relay-server/src/actors/healthcheck.rs

```diff
@@ -41,27 +59,47 @@ impl fmt::Display for SendError {
 
 impl Error for SendError {}
 
-/// Channel for sending public messages into a service.
+/// The address for a [`Service`].
 ///
-/// To send a message, use [`Addr::send`].
-#[derive(Clone, Debug)]
-pub struct Addr<T> {
-    tx: mpsc::UnboundedSender<Message<T>>,
+/// The address of a [`Service`] allows you to [send](Addr::send) messages to the service as
+/// long as the service is running.  It can be freely cloned.
+#[derive(Debug)]
+pub struct Addr<S: Service> {
+    tx: mpsc::UnboundedSender<S::Envelope>,
 }
 
-impl<T> Addr<T> {
+// Manually derive clone since we do not require `S: Clone` and the Clone derive adds this
+// constraint.
+impl<S: Service> Clone for Addr<S> {
+    fn clone(&self) -> Self {
+        Self {
+            tx: self.tx.clone(),
+        }
+    }
+}
+
+impl<S: Service> Addr<S> {
     /// Sends an asynchronous message to the service and waits for the response.
     ///
     /// The result of the message does not have to be awaited. The message will be delivered and
     /// handled regardless. The communication channel with the service is unbounded, so backlogs
     /// could occur when sending too many messages.
     ///
     /// Sending the message can fail with `Err(SendError)` if the service has shut down.
-    pub async fn send(&self, data: T) -> Result<bool, SendError> {
-        let (responder, rx) = oneshot::channel();
-        let message = Message { data, responder };
-        self.tx.send(message).map_err(|_| SendError)?;
-        rx.await.map_err(|_| SendError)
+    // Note: this is written as returning `impl Future` instead of `async fn` in order not
+    // to capture the lifetime of `&self` in the returned future.
+    pub fn send<M>(&self, message: M) -> impl Future<Output = Result<M::Response, SendError>>
+    where
+        M: ServiceMessage<S>,
+    {
+        let (envelope, response_rx) = message.into_envelope();
+        let res = self.tx.send(envelope).map_err(|_| SendError);
+        async move {
+            match res {
+                Ok(_) => response_rx.await.map_err(|_| SendError),
+                Err(_) => Err(SendError),
```

Comment:
The error here is already a `SendError`. There are two ways you could write this: 

```rust
        let res = self.tx.send(envelope).map_err(|_| SendError);
        async move {
            res?;
            response_rx.await.map_err(|_| SendError)
        }
```

or even better:

```rust
        let res = self.tx.send(envelope);
        async move {
            res.map_err(|_| SendError)?;
            response_rx.await.map_err(|_| SendError)
        }
```